File: questions/721.AccountsMerge.js
/**
 * accountsMerge
 *
 * @param {string[][]} accounts
 * @return {string[][]}
 */
const accountsMerge = function (accounts) {
  const users = new Set()
  const emailsMap = new Map()

  /**
   * @param {(string|Set)[]} from
   * @param {(string|Set)[]} to
   */
  const merge = (from, to) => {
    if (from === to) return

    users.delete(from)

    for (const email of from[1]) {
      emailsMap.set(email, to)
      to[1].add(email)
    }
  }

  for (const account of accounts) {
    const [name, ...emails] = account
    const user = [name, new Set()]

    users.add(user)
    for (const email of emails) {
      if (emailsMap.has(email)) {
        merge(emailsMap.get(email), user)
      } else {
        emailsMap.set(email, user)
        user[1].add(email)
      }
    }
  }

  const result = []

  for (const user of users) {
    const list = [...user[1]].sort()
    list.unshift(user[0])
    result.push(list)
  }

  return result
}
console.log(
  accountsMerge([
    ['John', 'johnsmith@mail.com', 'john_newyork@mail.com'],
    ['John', 'johnsmith@mail.com', 'john00@mail.com'],
    ['Mary', 'mary@mail.com'],
    ['John', 'johnnybravo@mail.com'],
  ]),
)

File: questions/721.AccountsMerge.go
package questions

import "sort"

type AccountsMergeUser struct {
	Name   string
	Emails map[string]string
}

func accountsMerge(accounts [][]string) [][]string {
	users := map[*AccountsMergeUser]struct{}{}
	emailMaps := map[string]*AccountsMergeUser{}
	var merge func(from, to *AccountsMergeUser)

	merge = func(from, to *AccountsMergeUser) {
		if from == to {
			return
		}

		delete(users, from)

		for email := range from.Emails {
			emailMaps[email] = to
			to.Emails[email] = ""
		}
	}

	for _, account := range accounts {
		name := account[0]
		emails := append([]string{}, account[1:]...)

		user := &AccountsMergeUser{
			Name:   name,
			Emails: map[string]string{},
		}

		for _, email := range emails {
			u, ok := emailMaps[email]
			if ok {
				merge(u, user)
			} else {
				emailMaps[email] = user
				user.Emails[email] = ""
			}
		}
		users[user] = struct{}{}
	}

	var result [][]string
	for user := range users {
		v := make([]string, len(user.Emails))

		i := 0
		for e := range user.Emails {
			v[i] = e
			i++
		}
		sort.Strings(v)
		v = append(v[:1], v[0:]...)
		v[0] = user.Name

		result = append(result, v)
	}

	return result
}
